src/banded/bidiagonalconjugationdata.jl

```diff
@@ -0,0 +1,85 @@
+"""
+    BidiagonalConjugationData{T, MU, MC} <: LazyMatrix{T}
+
+Struct for efficiently representing the matrix product `A = inv(U)XV`,
+assuming that 
+
+- `A` is upper bidiagonal, 
+- `U` is upper Hessenberg,
+- `X` is banded,
+- `V` is banded.
+
+None of these properties are checked internally. It is the user's responsibility
+to ensure these properties hold and that the product `inv(U)XV` is indeed bidiagonal.
+
+# Fields 
+- `U`: The upper Hessenberg matrix. 
+- `C`: The matrix product `XV`.
+- `dv`: A vector giving the diagonal of `A`. 
+- `ev`: A vector giving the superdiagonal of `A`.
+
+The vectors `dv` and `ev` grow on demand from `getindex` and should not be 
+used directly. Simply treat 
+
+    A = BidiagonalConjugationData(U, X, V)
+
+as you would an upper bidiagonal matrix.
+"""
+struct BidiagonalConjugationData{T,MU,MC} <: LazyMatrix{T}
+    U::MU
+    C::MC
+    dv::Vector{T}
+    ev::Vector{T}
+end
+function BidiagonalConjugationData(U::MU, X::MX, V::MV) where {MU,MX,MV}
+    C = X * V
+    T = promote_type(typeof(inv(U[begin])), eltype(U), eltype(C)) # include inv so that we can't get Ints
+    dv, ev = T[], T[]
+    return BidiagonalConjugationData{T,MU,typeof(C)}(U, C, dv, ev)
+end
+MemoryLayout(::Type{<:BidiagonalConjugationData}) = BidiagonalLayout{LazyLayout,LazyLayout}()
```

Comment:
We need to be careful: the interface for `BidiagonalLayout` includes overloading `diagonaldata` and `offdiagonaldata` but this isn't currently possible.

This is why I suggested a `BidiagonalConjugationData`/`BidiagonalConjugationBand` model.

Reply:
I might not be exactly understanding the `BidiagonalConjugationData`/`BidiagonalConjugationBand` model or the Lanczos example in your other comment. Are you suggesting something like: 

```julia
struct BidiagonalConjugationData{T,MU,MC}
    U::MU
    C::MC
    dv::Vector{T}
    ev::Vector{T}
    uplo::Char
end
function BidiagonalConjugationData(U::MU, X::MX, V::MV, uplo) where {MU,MX,MV}
    C = X * V
    T = promote_type(typeof(inv(U[begin])), eltype(U), eltype(C)) # include inv so that we can't get Ints
    dv, ev = T[], T[]
    return BidiagonalConjugationData{T,MU,typeof(C)}(U, C, dv, ev, _to_uplo(uplo))
end

struct BidiagonalConjugation{T, B <: BidiagonalConjugationData{T}} <: AbstractCachedMatrix{T}
    data::B
end
function BidiagonalConjugation(U, X, V, uplo)
    data = BidiagonalConjugationData(U, X, V, uplo)
    return BidiagonalConjugation(data)
end
```

Also, where is `offdiagonaldata` defined? I can't seem to locate it. Are you referring to the `(sub/sup)diagonaldata` pair?

Reply:
Actualy, I see exactly what you mean now. I'll get some commits in soon.